File: src/algorithms/python/Peeking_Iterator.py
# -*- coding: utf-8 -*-

from itertools import tee


class Iterator(object):
    def __init__(self, nums):
        """
        Initializes an iterator object to the beginning of a list.
        :type nums: List[int]
        """
        self.iterator = iter(nums)

    def hasNext(self):
        """
        Returns true if the iteration has more elements.
        :rtype: bool
        """
        try:
            self.iterator, _next = tee(self.iterator)
            _next.next()
        except:
            return False
        return True

    def next(self):
        """
        Returns the next element in the iteration.
        :rtype: int
        """
        return self.iterator.next()


class PeekingIterator(object):
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self.iterator = iterator
        self._tag = False
        self._lastnum = 0

    def peek(self):
        """
        Returns the next element in the iteration
        without advancing the iterator.
        :rtype: int
        """
        if not self._tag:
            self._tag = True
            self._lastnum = self.iterator.next()
        return self._lastnum

    def next(self):
        """
        :rtype: int
        """
        if not self._tag:
            self._lastnum = self.iterator.next()
        self._tag = False
        return self._lastnum

    def hasNext(self):
        """
        :rtype: bool
        """
        if not self._tag:
            return self.iterator.hasNext()
        else:
            return True


if __name__ == "__main__":
    iter = PeekingIterator(Iterator([1, 2, 3, 4]))
    while iter.hasNext():
        val = iter.peek()
        iter.next()

File: src/algorithms/cpp/Peeking_Iterator.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

// Below is the interface for Iterator, which is already defined for you.
// **DO NOT** modify the interface for Iterator.
class Iterator {
    struct Data;
	Data* data;
public:
	Iterator(const vector<int>& nums);
	Iterator(const Iterator& iter);
	virtual ~Iterator();
	// Returns the next element in the iteration.
	int next();
	// Returns true if the iteration has more elements.
	bool hasNext() const;
};

class PeekingIterator : public Iterator {
private:
    Iterator it_copy;
    int next_cache;
    bool flag;

public:
	PeekingIterator(const vector<int>& nums) : Iterator(nums), it_copy(Iterator(nums)), flag(false) {
	    // Initialize any member here.
	    // **DO NOT** save a copy of nums and manipulate it directly.
	    // You should only use the Iterator interface methods.
	}

    // Returns the next element in the iteration without advancing the iterator.
	int peek() {
        if (!flag) {
            next_cache = it_copy.next();
            flag = true;
        }
        return next_cache;
	}

	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	int next() {
        if (!flag)  it_copy.next();
        flag = false;
        return Iterator::next();
	}

	bool hasNext() const {
        return Iterator::hasNext();
	}
};

int main() {
    return 0;
}
